add quickstart docs from NuttX Companion (#1675)
apache/incubator-nuttx: Documentation/quickstart/quickstart.rst

.. include:: /substitutions.rst
.. _quickstart:

Quickstart
==========

> Here's the quick version of getting things going. This is a bare-bones outline for experienced developers– if it's
> going too quickly, dive into the other sections of the Companion. This Quickstart guide assumes you're on a Linux
> computer, you're using an ARM processor on your embedded board, and you're familiar with using the command line.

Reviewer: ```suggestion
Here's the quick version of getting started with NuttX. This is a bare-bones outline for experienced developers– if it's
going too quickly, dive into the following sections. This Quickstart guide assumes you're on a Linux
computer, you're using an ARM microcontroller on your embedded board, and you're familiar with using the command line.
```

#. Install a Cross-Compiler Toolchain

   With NuttX, you compile the operating system and your application on your desktop or laptop computer, then install the
   binary file on your embedded computer. This guide assumes your computer is an
   `ARM <https://en.wikipedia.org/wiki/ARM_architecture>`_ CPU. If it isn't, you'll need a different tool chain.

>    Download the right flavor of the
>    `ARM Embedded Gnu Toolchain <https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/gnu-rm>`_
>    for your embedded processor's CPU.

Reviewer: ```suggestion
   You can download a toolchain from
   `ARM Embedded Gnu Toolchain <https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/gnu-rm>`_
   for your embedded processor's CPU. You can also use a toolchain shipped with your OS for the `none-eabi` target, such as `gcc-arm-none-eabi` in Linux.
```

>    Unpack it into ``/opt/gcc`` and add the bin directory to your path. For instance:
> 
>    .. code-block:: bash
> 
>       $ sudo mkdir /opt/gcc
>       $ sudo chgrp -R users /opt/gcc
>       $ cd /opt/gcc
>       $ wget https://developer.arm.com/-/media/Files/downloads/gnu-rm/9-2019q4/gcc-arm-none-eabi-9-2019-q4-major-x86_64-linux.tar.bz2?revision=108bd959-44bd-4619-9c19-26187abf5225&la=en&hash=E788CE92E5DFD64B2A8C246BBA91A249CB8E2D2D
>       $ tar xf gcc-arm-none-eabi-9-2019-q4-major-x86_64-linux.tar.bz2
>       $ # add the toolchain bin/ dir to your path...
>       $ # you can edit your shell's rc files if you don't use bash
>       $ echo "export PATH=/opt/gcc/gcc-arm-none-eabi-9-2019-q4-major/bin:$PATH" >> ~/.bashrc

Reviewer: ```suggestion

In the following example, we download ``gcc-arm-none-eabi`` version 9.0 and unpack it into ``/opt/gcc`:

   .. code-block:: console

      $ sudo mkdir /opt/gcc
      $ sudo chgrp -R users /opt/gcc
      $ cd /opt/gcc
      $ wget https://developer.arm.com/-/media/Files/downloads/gnu-rm/9-2019q4/gcc-arm-none-eabi-9-2019-q4-major-x86_64-linux.tar.bz2
      $ tar xf gcc-arm-none-eabi-9-2019-q4-major-x86_64-linux.tar.bz2
      
Then, add the toolchain ``bin/`` directory to your path:

.. code-block:: console

      $ echo "export PATH=/opt/gcc/gcc-arm-none-eabi-9-2019-q4-major/bin:$PATH" >> ~/.bashrc

If you are using any other shell, the procedure is similar by editing the corresponding rc file.
```

#. Download Apache NuttX

> 

Reviewer: ```suggestion
The next step is to download NuttX main repository along the application repository. The latter is technically optional in a very minimal configurations but should be included in normal configuration since it includes the NuttX shell.
```

>    .. code-block:: bash

Reviewer: ```suggestion
   .. code-block:: console
```
Bash is for scripts, console is for sessions.

      $ mkdir nuttx
      $ cd nuttx
      $ git clone https://github.com/apache/incubator-nuttx.git nuttx
      $ git clone https://github.com/apache/incubator-nuttx-apps apps

>       $ git clone https://starcat-io@bitbucket.org/nuttx/tools.git tools

Reviewer: ```suggestion
      $ git clone https://bitbucket.org/nuttx/tools.git tools
```
We should try to get this into GitHub at least. It is strange to point to a bitbucket repo.

Author: @v01d That's possible, but under what account? It can't live in Apache's account because it contains non-Apache-License-compatible software. There was discussion about this on on the mailing list...

Reviewer: I think there's a `nuttx` organization. It could also be a `nuttx_extra` or `nuttx_contrib` if it sounds better. Anyway, just thinking out loud.

#. Install the ``kconfig-frontends`` package

>    This is necessary to use the menuconfig system to configure NuttX, and includes the ``kconfig-tweak`` utility
>    that can be used to quickly change debug settings.

Reviewer: ```suggestion
   NuttX is configured using ``kconfig`` system via an interactive menu system (``menuconfig``). It also includes the ``kconfig-tweak`` utility that can be used to quickly change debug settings without going into the menu system.
```

>    .. code-block:: bash

Reviewer: ```suggestion
   .. code-block:: console
```

>       $ cd tools/kconfig-frontends

Reviewer: On Ubuntu 20.04, you can simply install kconfig-frontends package. We should highlight the easy and common option first and then go into more involved situations like older Ubuntu versions or OSs.
We should also use tabbed examples for such scenarios (different instructions for each OS). Maybe you can add a "todo" for this.

Author: @v01d I'll add a todo for it. I'm still on 18.04, so I did this with what I have. Upgrading to 20.04 is on my personal todo list, but I'm hesitant to mess with a working development workstation...

Reviewer: We can consider latest LTS and non-LTS releases in instructions. I think that should cover most cases.
Anyway, you can just mention "apt install kconfig-frontends" for that case.

Author: @v01d I added a case for 20.04 as the first option.

      $ # on MacOS do the following:
      $ patch < ../kconfig-macos.diff -p 1
      $ ./configure --enable-mconf --disable-shared --enable-static --disable-gconf --disable-qconf --disable-nconf
      $ # on Linux do the following:
      $  ./configure --enable-mconf --disable-nconf --disable-gconf --disable-qconf
      $ make
      $ make install

#. List Possible Apache NuttX Base Configurations

   Find your hardware and a good starting application in the list of base configurations. In the application list,
   ``nsh`` is the Apache NuttX Shell, an interactive commandline that's a good starting place if you're new.

   .. code-block:: bash

      $ cd nuttx
      $ ./tools/configure.sh list | less

#. Initialize Configuration

   Pick one of the board:application base configuration pairs from the list, and feed it to
   the configuration script. The ``-l`` tells us that we're on Linux. macOS and Windows builds

>    are possible, this Companion doesn't cover them yet.

Reviewer: This mentions the "companion". Also, please add a todo for the missing OS instructions

Author: @v01d Will fix.

   .. code-block:: bash

      $ cd nuttx

>       $ ./tools/configure.sh -l <board-name>:<config-dir>

Reviewer: Maybe it would be good to use a real board as an example. I'm thinking a Nucleo which seems friendly and well supported.

Author: @v01d I have a Nucleo, and can look at that another time. For now I'll use the SAMA5D2-XULT which matches the rest of this section; we need to figure out how to add several other boards, maybe using tabs or something. I'll think about how to do this.

Author: @v01d I added an example. We can add more examples after we merge this PR.

    
#. Customize Your Configuration (Optional)

   This step is optional. Right now, this is mainly to get familiar with how it works– you don't need to change
   any of the options now, but knowing how to do this will come in handy later.

   There are a lot of options. We'll cover a few of them here. Don't worry about the complexity– you don't have to use most of the options.

   .. code-block:: bash

      $ make menuconfig

> #. Exit the menuconfig System
> 
>    Use the left or right arrow keys to select ``<Exit>``, and save the configuration.

Reviewer: ```suggestion
Use your arrows to navigate the menu and ``ENTER`` key to enable/disable options. To exit and save your configuration, go back to the main menu, choose ``<Exit>`` and select "yes" when asked if you want to save.
```

>    |br|

Reviewer: Any reason for this?

Author: For the `|br|`? If so, the spacing looked off without it. But can remove if you want.

Reviewer: Yes, please remove it. I'd rather fix appearance on CSS if needed.

#. Compile Apache NuttX

   .. code-block:: bash

>       $ make clean; make

Reviewer: ```suggestion
      $ make
```
I think it is confusing to require the clean right from the start.

#. Install the Executable Program on Your Board

>    This step is a bit more complicated, depending on your board. It's covered in the section

Reviewer: This is also a good place to have an easy to flash board such as a nucleo. A one line call into openocd is all it takes.

Author: @v01d Agreed that we need more detail on how to flash, using popular boards. But I don't want to do that as part of this PR. Let's save it for a future update.

Author: @v01d I have a Nucleo, and can work on this in my spare time in the next few weeks, but don't have time to do it as part of this PR.

Reviewer: Sure!

   for your embedded processor's CPU. You can also use a toolchain shipped with your OS for the `none-eabi` target, such as `gcc-arm-none-eabi` in Linux.

In the following example, we download ``gcc-arm-none-eabi`` version 9.0 and unpack it into ``/opt/gcc`:

   .. code-block:: console

      $ sudo mkdir /opt/gcc
      $ sudo chgrp -R users /opt/gcc
      $ cd /opt/gcc
      $ wget https://developer.arm.com/-/media/Files/downloads/gnu-rm/9-2019q4/gcc-arm-none-eabi-9-2019-q4-major-x86_64-linux.tar.bz2
      $ tar xf gcc-arm-none-eabi-9-2019-q4-major-x86_64-linux.tar.bz2
      
Then, add the toolchain ``bin/`` directory to your path:

.. code-block:: console

      $ echo "export PATH=/opt/gcc/gcc-arm-none-eabi-9-2019-q4-major/bin:$PATH" >> ~/.bashrc

If you are using any other shell, the procedure is similar by editing the corresponding rc file.

#. Download Apache NuttX
The next step is to download NuttX main repository along the application repository. The latter is technically optional in a very minimal configurations but should be included in normal configuration since it includes the NuttX shell.
   .. code-block:: console

      $ mkdir nuttx
      $ cd nuttx

>       $ git clone https://github.com/apache/incubator-nuttx.git nuttx

Reviewer: I'm thinking that maybe the quickstart should use stable release instead of cloning the repo. On the other hand, I think users are more accustomed to cloning with git. What do you think?

Author: @v01d Let's leave it as is, with cloning. We can add downloading a stable release later.

Author: @v01d @btashton The reason I think cloning is better right now is that there is no URL for the latest Apache NuttX release, for instance:

https://nuttx.apache.org/releases/latest/
https://downloads.apache.org/incubator/nuttx/latest/
https://downloads.apache.org/incubator/nuttx/latest/apache-nuttx-latest-incubating.tar.gz

(these links would always HTTP redirect to the latest release files...)

If we had that, we could put something in the docs that would not get out of date.

Reviewer 2: There are some Apache rules about pointing people at source instead of releases, but we can come back around to that.
My guess it we will end up having the link at /docs/latest point at the latest release and /docs/dev point at master which would then point at git with a banner that says "hey these are dev docs". 

The docs that we cut for a actual release 10.0.0 should certainly point at the release tarballs.

Author: @btashton We can do something in sphinx and add a tag that reads from a ``VERSION`` file or something, so as long as the version is updated, the docs will be correct.

Reviewer: Just a note: "latest" is used now to refer to master in the documentation. latest release would probably better named "stable".

Reviewer 2: Yeah we can figure the naming out later, I don't think this needs to hold this up, we just need to address at the least the release docs prior to the release.

Reviewer: Ok, we'll deal with this later. I'll resolve this conversation.

      $ git clone https://github.com/apache/incubator-nuttx-apps apps
      $ git clone https://bitbucket.org/nuttx/tools.git tools

#. Install the ``kconfig-frontends`` package

   NuttX is configured using ``kconfig`` system via an interactive menu system (``menuconfig``). It also includes the ``kconfig-tweak`` utility that can be used to quickly change debug settings without going into the menu system.

   .. code-block:: console

      $ cd tools/kconfig-frontends
      $ # on MacOS do the following:
      $ patch < ../kconfig-macos.diff -p 1
      $ ./configure --enable-mconf --disable-shared --enable-static --disable-gconf --disable-qconf --disable-nconf
      $ # on Linux do the following:
      $  ./configure --enable-mconf --disable-nconf --disable-gconf --disable-qconf
      $ make
      $ make install

#. List Possible Apache NuttX Base Configurations

   Find your hardware and a good starting application in the list of base configurations. In the application list,
   ``nsh`` is the Apache NuttX Shell, an interactive commandline that's a good starting place if you're new.

   .. code-block:: bash

      $ cd nuttx

>       $ ./tools/configure.sh list | less

Reviewer 3: `list` doesn't exist as an option, that's actually printing an error then dumping all the configurations.

Reviewer: You are right, there does not seem to be a "list" option.

Author: @Ouss4 @v01d Yes, I know I'm "misusing" the script here, but this does work, since if you put in an unknown option, it lists all the boards. Ideally we'd add an option to list the boards, but since this works now, is it ok to leave in? I could put a note or a todo. I'm also willing to add a `list` option to the script in another PR.

Reviewer: If you add a todo so that we remember to update it with the expected use after adding this feature to configure script I think it is ok to leave as is.

Reviewer 3: ```bash
diff --git a/tools/configure.sh b/tools/configure.sh
index bf6702d115..a16dadbcaa 100755
--- a/tools/configure.sh
+++ b/tools/configure.sh
@@ -58,6 +58,17 @@ Where:
 
 "
 
+function dumpcfgs
+{
+  configlist=`find ${TOPDIR}/boards -name defconfig`
+  for defconfig in ${configlist}; do
+    config=`dirname ${defconfig} | sed -e "s,${TOPDIR}/boards/,,g"`
+    boardname=`echo ${config} | cut -d'/' -f3`
+    configname=`echo ${config} | cut -d'/' -f5`
+    echo "  ${boardname}:${configname}"
+  done
+}
+
 # A list of optional files that may be installed
 
 OPTFILES="\
@@ -95,6 +106,10 @@ while [ ! -z "$1" ]; do
   -e )
     distclean=y
     ;;
+  list )
+    dumpcfgs
+    exit 0
+    ;;
   -h )
     echo "$USAGE"
     exit 0
@@ -134,13 +149,7 @@ if [ ! -d ${configpath} ]; then
     echo "Directory for ${boardconfig} does not exist.  Options are:"
     echo ""
     echo "Select one of the following options for <board-name>:"
-    configlist=`find ${TOPDIR}/boards -name defconfig`
-    for defconfig in ${configlist}; do
-      config=`dirname ${defconfig} | sed -e "s,${TOPDIR}/boards/,,g"`
-      boardname=`echo ${config} | cut -d'/' -f3`
-      configname=`echo ${config} | cut -d'/' -f5`
-      echo "  ${boardname}:${configname}"
-    done
+    dumpcfgs
     echo ""
     echo "$USAGE"
     exit 3
```
We can do this.  `configure.c` needs to be updated too, however. (I guess we can do that separately, there are other places where it's lagging behind, like custom boards support).

Reviewer 3: BTW, I agree on making the change on a separate PR.  Just one suggestion make it "--list" instead of just "list" as it is easier to adapt configure.c later.

Reviewer 3: I created #1697 for that.  Please take a look.

Reviewer 2: `-L` for listing is now merged.  Thanks @Ouss4

Author: Thank you @btashton @Ouss4 – I'll update the docs to use `-L` when I update this PR, should be able to do it tomorrow.

Author: @Ouss4 @btashton @v01d I updated the docs to use the new ``configure.sh -L`` option. 🙂 Thank you!
